src/library/scala/util/matching/Regex.scala

```diff
@@ -767,7 +767,37 @@ object Regex {
    *  }}}
    */
   object Groups {
-    def unapplySeq(m: Match): Option[Seq[String]] = if (m.groupCount > 0) Some(1 to m.groupCount map m.group) else None
+    def unapplySeq(m: Match): Option[Seq[String]] = {
+      if (m.groupCount > 0) extractGroupsFromMatch(m) else None
+    }
+  }
+
+  def extractGroupsFromMatch(m: Match): Option[List[String]] = {
```

Comment:
I cannot delegate the logic to extractGroupsFromMatcher(m.matcher) because if we called findAllInSource before, we will change the internal state of underlying matcher.

Reply:
I don't think that we need the if statement, it serves the same purpose as the existing while loop

Reply:
yes, sorry. Below is benchmark for latest code.

```
[info] Benchmark                                                                    Score     Error   Units
[info] RegexUnapplyBenchmark.t8022CharSequence                                    513.380 ?  15.950   ns/op
[info] RegexUnapplyBenchmark.t8022CharSequence:?gc.alloc.rate                     673.820 ?  20.141  MB/sec
[info] RegexUnapplyBenchmark.t8022CharSequence:?gc.alloc.rate.norm                544.001 ?   0.002    B/op
[info] RegexUnapplyBenchmark.t8022Match                                           641.295 ?  31.184   ns/op
[info] RegexUnapplyBenchmark.t8022Match:?gc.alloc.rate                           1443.362 ?  66.717  MB/sec
[info] RegexUnapplyBenchmark.t8022Match:?gc.alloc.rate.norm                      1453.690 ?  11.251    B/op
[info] RegexUnapplyBenchmark.t8787nullMatch                                         2.898 ?   0.031   ns/op
[info] RegexUnapplyBenchmark.t8787nullMatch:?gc.alloc.rate                          0.001 ?   0.002  MB/sec
[info] RegexUnapplyBenchmark.t8787nullMatch:?gc.alloc.rate.norm                    ? 10??              B/op
[info] RegexUnapplyBenchmark.t8787nullMatcher                                    2018.756 ?  62.720   ns/op
[info] RegexUnapplyBenchmark.t8787nullMatcher:?gc.alloc.rate                     1317.296 ?  50.693  MB/sec
[info] RegexUnapplyBenchmark.t8787nullMatcher:?gc.alloc.rate.norm                4180.003 ?  32.073    B/op

[info] Benchmark                                                                       (groupCount)     Score     Error   Units
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                              1   115.846 ?   6.062   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                               1  1231.906 ?  64.574  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                          1   224.000 ?   0.001    B/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                             10   734.242 ?  15.148   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                              10   831.198 ?  17.027  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                         10   960.001 ?   0.002    B/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark                                            100  8954.059 ? 100.950   ns/op
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate                             100   655.547 ?   7.191  MB/sec
[info] RegexUnapplyGroupsBenchmark.groupingBenchmark:?gc.alloc.rate.norm                        100  9240.015 ?   0.029    B/op
```